kernel/multicore_bringup/src/aarch64.rs

```diff
@@ -0,0 +1,214 @@
+    ap_mmu_config: Volatile<PhysicalAddress>,
+}
+
+pub struct MulticoreBringupInfo;
+
+pub fn handle_ap_cores(
+    kernel_mmi_ref: &MmiRef,
+    _multicore_info: MulticoreBringupInfo,
+) -> Result<usize, &'static str> {
+    let mut online_cores = 0;
+
+    // This ApTrampolineData & MmuConfig will be read and written-to
+    // by all detected CPU cores, via both its physical
+    // and virtual addresses.
+    let mmu_config = read_mmu_config();
+    let mut ap_data: ApTrampolineData = Default::default();
+    ap_data.ap_data_virt_addr.write(VirtualAddress::new_canonical(&ap_data as *const _ as usize));
+    ap_data.ap_stage_two_virt_addr.write(VirtualAddress::new_canonical(ap_stage_two as usize));
+
+    let entry_point_phys_addr: PhysicalAddress;
+    let ap_data_phys_addr: PhysicalAddress;
+    let mut ap_startup_mapped_pages: MappedPages; // must be held throughout APs being booted up
+    {
+        let mut kernel_mmi = kernel_mmi_ref.lock();
+        let page_table = &mut kernel_mmi.page_table;
+
+        // get the physical address of ap_entry_point
+        let entry_point_virt_addr = VirtualAddress::new_canonical(ap_entry_point as usize);
+        entry_point_phys_addr = page_table.translate(entry_point_virt_addr).unwrap();
+
+        // get the physical address of the MmuConfig
+        let mmu_config_virt_addr = VirtualAddress::new_canonical(&mmu_config as *const _ as usize);
+        ap_data.ap_mmu_config.write(page_table.translate(mmu_config_virt_addr).unwrap());
+
+        // Note: this could be variable on aarch64
+        // we'd need a way to find any identity mappable page+frame range
+        const AP_STARTUP: usize = 0x10000;
+
+        // When the AArch64 core will enter startup code, it will do so with MMU disabled,
+        // which means these frames MUST be identity mapped.
+        let ap_startup_frames = memory::allocate_frames_by_bytes_at(PhysicalAddress::new_canonical(AP_STARTUP), PAGE_SIZE)
+            .map_err(|_e| "handle_ap_cores(): failed to allocate AP startup frames")?;
+        let ap_startup_pages  = memory::allocate_pages_at(VirtualAddress::new_canonical(AP_STARTUP), ap_startup_frames.size_in_frames())
+            .map_err(|_e| "handle_ap_cores(): failed to allocate AP startup pages")?;
+
+        // map this RWX
+        let flags = PteFlags::new().valid(true).writable(true);
+        ap_startup_mapped_pages = page_table.map_allocated_pages_to(
+            ap_startup_pages,
+            ap_startup_frames,
+            flags,
+        )?;
+
+        // get physical address of the ApTrampolineData structure
+        ap_data_phys_addr = page_table.translate(ap_data.ap_data_virt_addr.read()).unwrap();
+
+        // copy the entry point code
+        let dst = ap_startup_mapped_pages.as_slice_mut(0, PAGE_SIZE).unwrap();
+        let src = unsafe { (ap_entry_point as *const [u8; PAGE_SIZE]).as_ref() }.unwrap();
+        dst.copy_from_slice(src);
```

Comment:
same question for this -- what are the requirements here? This whole complex concept of copying code into a dedicated area is just to ensure that everything can be accessed in x86 real mode (16-bit addressing, no paging, etc)